### ISSUE TITLE
Implement Mini-Batch Training

### ISSUE THREAD

**Issue by jocooler** [cazala/synaptic#126]
It's possible I just missed the code for this in the source, if so, I apologize. Also, I'm very new to neural networks, but after implementing this, my training seemed to be more effective.

To quote from [Michael Nielsen's book](http://neuralnetworksanddeeplearning.com/chap1.html):

> We can think of stochastic gradient descent as being like political polling: it's much easier to sample a small mini-batch than it is to apply gradient descent to the full batch, just as carrying out a poll is easier than running a full election. For example, if we have a training set of size n=60,000, as in MNIST, and choose a mini-batch size of (say) m=10, this means we'll get a factor of 6,000 speedup in estimating the gradient! Of course, the estimate won't be perfect - there will be statistical fluctuations - but it doesn't need to be perfect: all we really care about is moving in a general direction...

This seems pretty straightforward to implement, and since it can really have an impact on training, perhaps it should be included in the library. For my training sets, it really speeds up training and improves results.

```
/* Usage:
miniBatchTrain(10000, trainer, trainingSet, 20, {iterations: 1, rate: 0.01}); // set iterations to 1 so we just run each minibatch once, and the learning rate small to minimize learning from wonky values, 
*/ 

function miniBatchTrain(iterations, trainer, trainingSet, MiniBatchSize, options){
  var i = 0;
  while (i < iterations) {
    var workingTrainingSet = trainingSet.slice(0);
    shuffle(workingTrainingSet);
    
    while (workingTrainingSet.length > MiniBatchSize) {
      var these = workingTrainingSet.splice(0, MiniBatchSize);
      trainer.train(these, options);
    }
    i++;
  }
}

/* Shuffle is a generic shuffle algorithm, I just modified this one from SO */
/**
 * Shuffles copy of an array.
 * From http://stackoverflow.com/questions/6274339/how-can-i-shuffle-an-array-in-javascript
 * @param {Array} a items The array containing the items.
 */
function shuffle(a) {
    var j, x, i, b=a;
    for (i = b.length; i; i--) {
        j = Math.floor(Math.random() * i);
        x = b[i - 1];
        b[i - 1] = b[j];
        b[j] = x;
    }
    return b;
}
```
